Add codes to generate csv files for hunter and update README

marianne-manaog opened the pull request:
These codes create one csv file for each test type (e.g., `lwt-fixed-100-partitions`,  `lwt-fixed-1000-partitions`, etc.). It also adds unit tests for utility functions used to create the csv files for hunter.

mfleming commented on README.md:
```diff
@@ -5,12 +5,21 @@
+This repository contains:
+- performance tests for Cassandra 
 databases using the [Fallout tool](https://github.com/datastax/fallout).
+- codes to create the csv file that will be fed to hunter.
```
Can you separate out the running of fallout and fallout+hunter? Some users will only want to run the tests with fallout but not use hunter.

marianne-manaog replied:
Thanks, addressed at [87fcf9d](https://github.com/datastax/fallout-tests/pull/5/commits/87fcf9d6bb3b4aba5d5544739b9832fe4925feea)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+    input_df.to_csv(path_to_output, index=False)
+
+
+def get_paths_to_six_json(path_w_spec_date: str) -> List[str]:
```
```suggestion
def get_paths_to_json(path_w_spec_date: str) -> List[str]:
```

marianne-manaog replied:
Thanks, addressed at [5586e5a](https://github.com/datastax/fallout-tests/pull/5/commits/5586e5a47d626f037ce9d085ddaa54d24fef1d13)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+
+def get_paths_to_six_json(path_w_spec_date: str) -> List[str]:
+    """
+    Get the paths to the six json files of the performance results, i.e., one for each
```
```suggestion
    Get the paths to the json files of the performance results, i.e., one for each
```

marianne-manaog replied:
Thanks, addressed at [5586e5a](https://github.com/datastax/fallout-tests/pull/5/commits/5586e5a47d626f037ce9d085ddaa54d24fef1d13)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+                        The test run's date-related path to the json file.
+
+    Returns:
+            A list of paths (List[str]) to the six json files of interest.
```
```suggestion
            A list of paths (List[str]) to the json files of interest.
```

marianne-manaog replied:
Thanks, addressed at [5586e5a](https://github.com/datastax/fallout-tests/pull/5/commits/5586e5a47d626f037ce9d085ddaa54d24fef1d13)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+                    date_sorted_y_d_m, is_prospective)
+
+        # cd back into hunter_csv
+        cd_into_proj_dir()
```
Can we make this stateless? e.g. return the full path the project directory as a string instead of altering the current working directory.

marianne-manaog replied:
Thanks, addressed at [1fa5cd9](https://github.com/datastax/fallout-tests/pull/5/commits/1fa5cd92044f07e0c33fa602f149cca23de4171c). I removed the function `cd_into_proj_dir` and the `GitPython` dependency, as I managed to use `git -C <PATH_TO_REPO> pull origin <MAIN_BRANCH_NAME>` via `subprocess` (running it as a bash command).

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+        )
+        return pd.DataFrame(), ''
+
+    test_type = json_paths[0].split(os.sep)[5]
```
Can you either calculate the index value or use a descriptive variable name that explains which part of the path we're indexing?

marianne-manaog replied:
Thanks, addressed at [a88cbe0](https://github.com/datastax/fallout-tests/pull/5/commits/a88cbe06fe955d66014ab5b995bf28e000c409ed)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+        # cd into filename with a date-specific test run (path_w_date)
+        cd_into_proj_dir(path_w_date)
+
+        path_to_each_test_json = get_paths_to_six_json(path_w_date)
```
```suggestion
        path_to_each_test_json = get_paths_to_json(path_w_date)
```

marianne-manaog replied:
Thanks, addressed at [5586e5a](https://github.com/datastax/fallout-tests/pull/5/commits/5586e5a47d626f037ce9d085ddaa54d24fef1d13)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+            # cd into filename with a date-specific test run (path_w_date)
+            cd_into_proj_dir(path_w_date)
+
+            path_to_each_test_json = get_paths_to_six_json(path_w_date)
```
```suggestion
            path_to_each_test_json = get_paths_to_json(path_w_date)
```

marianne-manaog replied:
Thanks, addressed at [5586e5a](https://github.com/datastax/fallout-tests/pull/5/commits/5586e5a47d626f037ce9d085ddaa54d24fef1d13)

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,477 @@
+        list_of_hunter_df = []
+        list_of_type_of_tests = []
+        for test_json_path in path_to_each_test_json:
+
```
Can you delete the empty line after `if` and `for`, just for consistency across the file.

marianne-manaog replied:
Thanks, addressed at [9b4c471](https://github.com/datastax/fallout-tests/pull/5/commits/9b4c471dc45999ae2509ca86b97e383517f57e39)

marianne-manaog commented:
Hi @mfleming, as per our offline discussion, I updated the logic to read the Git shas (for the `cassandra` and `fallout-tests` repos) from their respective log files instead of fetching them programmatically, and simplified other code logic too. Thank you for your feedback, which was useful to optimize these codes.

mfleming commented on README.md:
```diff
@@ -1,18 +1,25 @@
+
 
-## What is Fallout
+## Instructions to run performance tests via Fallout on Cassandra
```
```suggestion
## Running performance tests via Fallout on Cassandra
```

mfleming replied:
This is a minor change but I find it helpful to avoid including words that you can surmise from context, e.g. most things in this file are instructions.

mfleming commented on README.md:
````diff
@@ -38,3 +45,22 @@ output-dir             Where to write test run artifacts;
 ```
+
+
+## Instructions to create a csv file with results from performance tests for hunter
````
```suggestion
## Analysing performance test results with hunter
```

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,331 @@
+    nightly_result_dates.sort()
+
+    # Set to False if running this for the first time, then re-run and set to True.
+    is_case_prospective = PROSPECTIVE_MODE
```
In general, it's too high of a bar to expect users to change the code to alter the behaviour of an app/script. In 6 months someone is going to run this script and, I suspect, not read the code :) This should really be a command-line argument with the default being prospective.

Also, the explanation of prospective/retrospective should be in the README.

But these are minor suggestions and I think we should just create a follow up GH issue to record these two points and address it later. It doesn't need to hold up merging this PR.

marianne-manaog replied:
I have added the explanation on README at [99de7fd](https://github.com/datastax/fallout-tests/pull/5/commits/99de7fd8c3562ee2cc17c780c309c98e6f257c64) and as for the command line argument, will create a GH issue for this.

mfleming commented on src/scripts/create_hunter_csv.py:
```diff
@@ -0,0 +1,331 @@
+                    test_json_path)
+                if type_of_test:
+                    types_of_tests.append(type_of_test)
+                if SUBSTR_TESTS_NAMES[0] in type_of_test and not hunter_df.empty:
```
I like this change, it's almost what I expected but can't you iterate over the SUBSTR_TESTS_NAMES and collapse all these if/elif into one now we have this dict? e.g.

```
for t in SUBSTR_TESTS_NAME:
   if t in type_of_test and not hunter_df.empty:
       SUBSTR_TESTS_NAME[t].append(hunter_df)
    else:
        get_error_log(type_of_test)
```

?

marianne-manaog replied:
I tried running the code above but I got: 
`File "create_hunter_csv.py", line 287, in <module>
    SUBSTR_TESTS_NAMES[t].append(hunter_df)
TypeError: list indices must be integers or slices, not str`. 

And also, tried:
```
for i, t in enumerate(SUBSTR_TESTS_NAMES):
     if t in type_of_test and not hunter_df.empty:
           SUBSTR_TESTS_NAMES[i].append(hunter_df)
     else:
           get_error_log(type_of_test)
```

But still erroring. As you have said we can just create a follow up GH issue to record these two points and address it later.

mfleming reviewed (approved):
Looks good, approved!  I left a few minor comments (to be addressed in a follow up ticket) and one question on using the `SUBSTR_TESTS_NAMES` dict to reduce the amount of code in the `if/elif` block -- if you can use my suggestion great! But if for some reason it doesn't work (my Python fu is weak) then just merge this as-is.
